### DISCUSSION TITLE
comparison results between gramine and scone

### DISCUSSION THREAD

**Discussion by jyizheng** [gramineproject/gramine#1489] [General]
Hi Gramine experts,

I am wondering what are the differences between Gramine and Scone. It seems Gramine also implements techniques of Scone (exitless calls). What is the relative performance b/w  the two? Are there any up-to-date results that I can refer to?

Thanks,

**Comment by kailun-qin** [comment]
Hi @jyizheng! Pls kindly see my comments below.

> I am wondering what are the differences between Gramine and Scone.

Googling around, there was one paper: https://sys.cs.fau.de/extern/lehre/ws22/akss/material/gramine-scone.pdf making this direct comparision. I hope it provides some insights, but w/o a thorough read, I can't tell whether the comparisions/results are all rigorous and up-to-date. Actually, there are a host of papers discussing differences (of various aspects) between "TEE containers" incl., Gramine/SCONE.

> It seems Gramine also implements techniques of Scone (exitless calls).

Yes, Gramine does support the `Exitless (or Switchless) calls`. For details, pls see: https://gramine.readthedocs.io/en/latest/performance.html#exitless-feature.

Note that there are in general two types of implementations for exitless calls: `busy-wait` and `sleep-wakeup` (depending on how the caller thread acts). Gramine adopted the `busy-wait` approach while SCONE uses the `sleep-wakeup` approach IIRC.

> What is the relative performance b/w the two? Are there any up-to-date results that I can refer to?

No, I'm not sure if there is a public and up-to-date one. Since both projects are still evolving, I think the best way to get the up-to-date results is to benchmark against their latest versions.
